src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,210 @@
+# 😱 Status quo stories: Barbara carefully dismisses embedded `Future`
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!
+
+## The story
+
+[Barbara] is contributing to an [OS](https://www.tockos.org) that supports
+running multiple applications on a single microcontroller. These
+microcontrollers have as little as 10's of kilobytes of RAM and 100's of
+kilobytes of flash memory for code. Barbara is writing a library that is used by
+multiple applications -- and is linked into each application -- so the library
+is very resource constrained. The library should support asynchronous operation,
+so that multiple APIs can be used in parallel within each (single-threaded)
+application.
+
+Barbara begins writing the library by trying to write a console interface, which
+allows byte sequences to be printed to the system console. Here is an example
+sequence of events for a console print:
+
+1. The interface gives the kernel a callback to call when the print finishes,
+   and gives the kernel the buffer to print.
+1. The kernel prints the buffer in the background while the app is free to do
+   other things.
+1. The print finishes.
+1. The app tells the kernel it is ready for the callback to be invoked, and the
+   kernel invokes the callback.
+
+Barbara tries to implement the API using
+[`core::future::Future`](https://doc.rust-lang.org/stable/core/future/trait.Future.html)
+so that the library can be compatible with the async Rust ecosystem. To do
+this, she decides to make the buffer printing API return a Future. She starts
+with a skeleton:
+
+```rust
+/// Passes `buffer` to the kernel, and prints it to the console. Returns a
+/// future that returns `buffer` when the print is complete. The caller must
+/// call kernel_ready_for_callbacks() when it is ready for the future to return. 
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
```

Comment:
Is the reason Barbara did not use async/await notation relevant for this story?

Reply:
Yes, I think it is. My understanding is that async/await notation can only block on other futures, and that at the leaf nodes where Rust's async APIs meet an external async API you need handwritten futures. In this case, this driver was interacting directly with the kernel.

I added a sentence to the above paragraph ("Barbara tries to implement... skeleton") explaining this.